### PR TITLE
add VAETrainer for VAE sample

### PR DESCRIPTION
**Pull request by sugiyama34**
# Pull Request

## 内容

#192 
VAE sample の VAETrainer を追加

## 他のコード・機能への影響

- [x] なし
- [ ] あり

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点を記述、リストアップしましたか?
- [x] PR を`make run`コマンドでローカルでテストしましたか?

## 補足

- l93 の `dataset = TensorDataset(torch.stack(list(data)))` に変更を入れました。
  - もともと `torch.stack()` に `data` を入れていましたが、 `list(data)` を入れています。
  本当はちゃんと `data` は list なので不要ですが、`make run` の formatter に怒られたので、明示的に list になるように修正しました。
- reconstruction loss (`recon_loss`) は、MSE の半分にしてあります
  - こうすると、Decoder が出す確率分布が、分散1の正規分布である場合に対応します

### CODE REVIEW

File: samples/vae-torch/trainer.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        for epoch in range(self.max_epochs):
+            recon_sum = 0.0
+            kl_sum = 0.0
+            data_count = 0
+            for (batch,) in dataloader:
+                batch = batch.to(device)
+
+                # Forward pass
+                dist = self.encoder(batch)
+                z = self.decoder(dist.rsample())
```
[the remark this answers is not in the document]
Comment (sugiyama34) [reply samples/vae-torch/trainer.py:110]: ほんとだ、間違ってる、、、！

File: samples/vae-torch/trainer.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    def create_optimizers(self) -> OptimizersSetup:
+        params = itertools.chain(
+            self.encoder.model.parameters(),
+            self.decoder.model.parameters(),
+        )
+        """Create optimizers for the encoder and decoder models.
+
+        Returns:
+            OptimizersSetup: A dictionary containing the optimizer for the VAE.
+        """
```
[the remark this answers is not in the document]
Comment (sugiyama34) [reply samples/vae-torch/trainer.py:79]: ほんまや、なんでやねん。
コピペミスったっぽいです。すみません。

File: samples/vae-torch/trainer.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                )
+                self.writer.add_scalar(
+                    "Loss/Reconstruction/Batch",
+                    recon_loss.item() / batch_size,
+                    batch_steps,
+                )
+                self.writer.add_scalar(
+                    "Loss/KL/Batch", kl_loss.item() / batch_size, batch_steps
+                )
+                batch_steps += 1
```
Comment (Geson-anko) [inline samples/vae-torch/trainer.py:151]: これは `self.global_steps`を定義して、それを利用してログしてください。

```suggestion
                self.writer.add_scalar(
                    "Loss/Total/Batch", loss.item() / batch_size, self.global_step
                )
                self.writer.add_scalar(
                    "Loss/Reconstruction/Batch",
                    recon_loss.item() / batch_size,
                    self.global_step,
                )
                self.writer.add_scalar(
                    "Loss/KL/Batch", kl_loss.item() / batch_size, self.global_step
                )
                self.global_step+= 1
```

File: samples/vae-torch/trainer.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+            # accumulate metrics per epoch
+            avg_total = (recon_sum + kl_sum) / data_count
+            avg_recon = recon_sum / data_count
+            avg_kl = kl_sum / data_count
+
+            # log per epoch
+            self.writer.add_scalar("Loss/Total/Epoch", avg_total, epoch)
+            self.writer.add_scalar("Loss/Reconstruction/Epoch", avg_recon, epoch)
+            self.writer.add_scalar("Loss/KL/Epoch", avg_kl, epoch)
```
Comment (Geson-anko) [inline samples/vae-torch/trainer.py:161]: これも `self.current_epoch`を定義して、それを使うと良いと思います。
```suggestion
            self.writer.add_scalar("Loss/Total/Epoch", avg_total, self.current_epoch)
            self.writer.add_scalar("Loss/Reconstruction/Epoch", avg_recon, self.current_epoch)
            self.writer.add_scalar("Loss/KL/Epoch", avg_kl, self.current_epoch)
```

File: samples/vae-torch/trainer.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+        # close writer after training
+        self.writer.close()
```
Comment (Geson-anko) [inline samples/vae-torch/trainer.py:164]: まだまだ学習は続くから closeしないで～

```suggestion
```

File: samples/vae-torch/trainer.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+        super().__init__(
+            training_condition_data_user="observation",
+            min_buffer_size=batch_size,
+            min_new_data_count=batch_size,
+        )
+        self.max_epochs = max_epochs
+        self.batch_size = batch_size
+        self.lr = lr
+
```
Comment (Geson-anko) [inline samples/vae-torch/trainer.py:51]:
```suggestion
        self.lr = lr
        self.global_step = 0
        self.current_epoch = 0

```

File: samples/vae-torch/trainer.py
Comment (Geson-anko) [inline samples/vae-torch/trainer.py:1]: Trainerの内部状態である `global_step`や `current_epoch`は、 `save_state` および `load_state`メソッドをoverrideして、適当に保存すると良いでしょう。

```py
    def save_state(self, path: Path) -> None:
       super().save_state(path)
       (path / "global_step).write_text(str(self.global_step))
       # current epochについても同様

    def load_state_(self, path: Path) -> None:
        super().load_state(path)
        self.global_step = int(...)
```

**Review by Geson-anko — CHANGES_REQUESTED** [review]
@sugiyama34 実装ありがとうございます。内部状態の取り扱いなどに少し修正があるのでよろしくお願いします

File: samples/vae-torch/trainer.py
Comment (sugiyama34) [reply samples/vae-torch/trainer.py:1]: 修正しました！
